internal/Yesod/Session/Persist/Prelude/Embedding.hs

```diff
@@ -0,0 +1,41 @@
+module Yesod.Session.Persist.Prelude.Embedding
+  ( Embedding (..)
+  , embed
+  , extract
+  , extractIgnoringError
+  , dimapEmbedding
+  ) where
+
+import Yesod.Session.Persist.Prelude.Reexports
+
+-- | Targets a value that is optionally present in some stateful monadic context
+data Embedding (con :: (Type -> Type) -> Constraint) e a = Embedding
+  { embed :: forall m. con m => Maybe a -> m ()
+  -- ^ Sets or clears the value
+  , extract :: forall m. (Functor m, con m) => m (Either e (Maybe a))
+  -- ^ Removes the value if present, returning what was removed
+  }
+
+embed :: con m => Embedding con e a -> Maybe a -> m ()
+embed Embedding {embed = x} = x
+
+extract :: (Functor m, con m) => Embedding con e a -> m (Either e (Maybe a))
+extract Embedding {extract = x} = x
+
+extractIgnoringError :: (Functor m, con m) => Embedding con e a -> m (Maybe a)
+extractIgnoringError e = extract e <&> fromRight Nothing
+
+dimapEmbedding
```

Comment:
Why not `bimapEmbedding`?

Reply:
Sort of a convention from [profunctors](https://hackage.haskell.org/package/profunctors-5.6.2/docs/Data-Profunctor.html). Latin "bi" is used for bifunctors, Greek "di" for profunctors.

Reply:
TIL